### PR TITLE
feat(@angular/cli): use separate tsconfigs

### PR DESCRIPTION
**Pull request by filipesilva**
This PR adds tsconfigs for each separate application:
- `src/tsconfig.app.json`: configuration for the Angular app.
- `src/tsconfig.spec.json`: configuration for the unit tests. Defaults to the Angular app config.
- `e2e/tsconfig.e2e.json`: configuration for the e2e tests.

There is an additional root-level `tsconfig.json` that is used for editor integration.

For new Angular version 4 projects, these tsconfigs will use inheritance since it's available with TypeScript 2.1.

This is not a breaking change. Existing projects should not be affected.

### CODE REVIEW

**Comment by deebloo** [comment]
fiesta! 💃

**Comment by filipesilva** [comment]
Olé! 💃

**Comment by hansl** [comment]
LGTM 💃

File: packages/@angular/cli/blueprints/ng2/files/__path__/tsconfig.app.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    "experimentalDecorators": true,
+    "lib": [
+      "es2016",
+      "dom"
+    ],<% } %>
+    "outDir": "<%= relativeRootPath %>/out-tsc/app",
+    "target": "es5",
+    "module": "es2015",
+    "baseUrl": "",
+    "types": []
```
Comment (dzonatan) [inline packages/@angular/cli/blueprints/ng2/files/__path__/tsconfig.app.json]: @filipesilva A bit late question but why declaring `types` manually is better than having one global `typeRoots` (as there was before)?
```json
"typeRoots": [
    "../node_modules/@types"
]
```

Comment (filipesilva) [inline packages/@angular/cli/blueprints/ng2/files/__path__/tsconfig.app.json]: A couple of reasons:
- Some typings might actually have duplicate indentifiers. This has happened before with `protractor` and `jquery` both using `$`.
- You might be using typings that really don't shouldn't in your app, like `describe` that should only be in unit tests, or node typings.

That being said, it is more convenient to use `typeRoots` if you're not too concerned with those things.

Comment (dzonatan) [inline packages/@angular/cli/blueprints/ng2/files/__path__/tsconfig.app.json]: Sounds convincing! Thanks for such a detail comment!! 🍺 🍺

**Comment by jtsom** [comment]
Just want to leave a comment about my issue - the statement above that "existing projects should not be affected" isn't entirely true...

**Comment by filipesilva** [comment]
@jtsom that was a mistake on our part. It's being fixed in https://github.com/angular/angular-cli/pull/5060

**Comment by filipesilva** [comment]
That's likely because you have to add it to the `types` array in `src/tsconfig.app.json`. We updated https://github.com/angular/angular-cli/wiki/stories-third-party-lib to also mention that.

**Comment by jtsom** [comment]
Was just about to make a comment about that too... would be nice if there was a mention in the rc.0 update docs to watch out for third party libs... Thanks

**Comment by filipesilva** [comment]
@jtsom adding in https://github.com/angular/angular-cli/pull/5147.

File: packages/@angular/cli/blueprints/ng2/files/__path__/tsconfig.app.json (hunk 1 above)
Comment (larssn) [reply packages/@angular/cli/blueprints/ng2/files/__path__/tsconfig.app.json]: This was hard to dig out, but thanks for clarifying. 👍

**Pull request merged**